linkerd/app/gateway/src/lib.rs

```diff
@@ -85,161 +93,246 @@ where
         dispatch_timeout,
         ..
     } = inbound.config().proxy.clone();
+    let local_id = inbound.runtime().identity.as_ref().map(|l| l.id().clone());
+
+    // For each gatewayed connection that is *not* HTTP, use the target from the
+    // transport header to lookup a service profile. If the profile includes a
+    // resolvable service name, then continue with TCP endpoint resolution,
+    // balancing, and forwarding. An invalid original destination address is
+    // used so that service discovery is *required* to provide a valid endpoint.
+    //
+    // TODO: We should use another target type that actually reflects
+    // reality. But the outbound stack is currently pretty tightly
+    // coupled to its target types.
+    let tcp = outbound
+        .clone()
+        .push_tcp_endpoint()
+        .push_tcp_logical(resolve.clone())
+        .into_stack()
+        .push_request_filter(|(p, _): (Option<profiles::Receiver>, _)| match p {
+            Some(rx) if rx.borrow().name.is_some() => Ok(outbound::tcp::Logical {
+                profile: Some(rx),
+                orig_dst: std::net::SocketAddr::from(([0, 0, 0, 0], 0)),
+                protocol: (),
+            }),
+            _ => Err(discovery_rejected()),
+        })
+        .push(profiles::discover::layer(profiles.clone(), {
+            let allow = allow_discovery.clone();
+            move |addr: NameAddr| {
+                if allow.matches(addr.name()) {
+                    Ok(addr)
+                } else {
+                    Err(RefusedNotResolved(addr))
+                }
+            }
+        }))
```

Comment:
nit/TIOLI: seems like there _could_ maybe be some kind of
```rust
impl NameMatch {
    // ...
    pub fn into_predicate<E: Error + Default>(&self) -> NamePredicate<E> {
        NamePredicate {
            m: self.clone(), _e: PhantomData,
        }
    }
}

struct NamePredicate<E: Default> {
    m: NameMatch,
    _e: PhantomData<fn(E)>,
}

impl<E, P> Predicate<E, P> for NamePredicate<E>
where
    E: Error + Default,
    P: Param<NameAddr>,
{
    type Request = P;
    fn check(&mut self, target: P) -> Result<Self::Request, Box<dyn Error + Send + Sync + 'static>> {
        if self.m.matches(target.param()) {
            Ok(target)
        } else {
            Err(E::default().into())
        }
    }
}
```
that could be used both here and in the http stack below?

Not sure if this is worth the effort.